Standardisation

tmleafs opened the pull request:
I have updated all DH's to match all the changes veeceeoh made to the Aqara temp this includes

- Moved the `preferences{} `section down to the bottom of` metadata{}`, below `tiles{}`, following the convention of the SmartThings Developer Documentation
- removed RequiredAtSetup from preferences{} section (it's only needed for SmartApps)
- removed displayDuringSetup: false because they aren't necessary
- Completely reorganized the preference page into distinct groups, with some reordering, and also changes to the text to remove redundant explanations and make them more concise. (screenshots below)
- lastCheckIn updates are no longer listed in the Events log ("Recently" tab of the ST mobile app) 
- Added code so that when the sensor is first paired, the Battery Changed is set to the current date
- I have renamed "Last Checkin" tile label to "Last Event"

bspranger commented:
Have you tested this in the IDE to ensure we don't break something? 

If so, I will merge

tmleafs commented:
I have tested original button and original & Aqara door as thats all I have.

alecm commented:
Are we removing the last checkin from the 'recently'  because it's duplicative of battery reporting?   I'll admit that I'm used to looking at that to make sure, say, a motion sensor, is still connected even when there's no motion around (so looking at a report come in every 55 minutes).  If there is *always* a battery report at regular check in cycle  (whatever timing for the particular device) (when there's no motion or anything else to trigger) then I'm ok w/ removing it.

alecm commented:
Are we removing the last checkin from the 'recently' because it's duplicative of battery reporting? I'll admit that I'm used to looking at that to make sure, say, a motion sensor, is still connected even when there's no motion around (so looking at a report come in every 55 minutes). If there is always a battery report at regular check in cycle (whatever timing for the particular device when there's no motion or anything else to trigger) then I'm ok w/ removing it.

tmleafs commented:
@alecm its only been hidden, will still show on the device event log on the IDE and in last event or as you said the last battery report in the recently log

veeceeoh commented:
Just took a quick peek at a couple of the DTHs - looks great so far, I'll check. more thoroughly when I get home in a while.

As mentioned in my (hopefully) last commit update to my pull req for the T/H sensor, I'm wondering if `section {}` in `preferences{}` has any effect when viewing the settings page of devices on Android devices, because it does absolutely nothing in iOS. From the Developer Docs, I gather that it's only used in SmartApps, but I want to be sure.

If section {} doesn't have any effect in DTHs on Android also, then I say get rid of it from the code...

veeceeoh commented:
@alecm - by design, the lastCheckin event only happens when there's a report message from the sensor (i.e.., battery, temp, humidity, pressure, motion detected/no motion , lux, open/close state change, wet/dry state change, and button pushed.)

In my updated code for the T/H sensor, I turned off the display of lastCheckin event messages for _everything_ except rmessages that can't be parsed. And I have not seen a single checkin event displayed since then. On my new motion detector, when I'm not at home and no motion is detected, every single lastCheckin happens at the same time as a battery report. Same thing for all my door/window sensors and my leak sensor.

So, these sensors _do not_ actually send a real check-in message. Users should know that during times of inactivity, battery reports are the best indicator that the sensor is still working and connected. I have seen a few instances of skipped battery reporting, but in those cases there's no checkin message. This means all the checkin message is doing is cluttering up the events log as seen in the "Recently" tab of the St mobile app.

However, I think having the last checkin date/time displayed in a tile is great, because if it's more than two hours ago, you need to see what's going on. So all types of reports in the new T/H sensor code still log an event, it's just not displayed. You can still find the full list of lastCheckin events by looking at the device in IDE, and click on lastCheckin in the Current State section. And these will of course all correspond to other report events.

I also believe the Last Checkin tile should be renamed to something like "Last report received" because that is more representative of what it's displaying.

alecm commented:
Thanks for the great explanation. But even so, given the challenges of the
Xiaomis (we're basically jailbreaking them) - I find it really useful as an
end user to not have to go to the IDE and just go to my app/ recent events
and see that it's checking in every 55 or whatever minutes to know that the
connection is still healthy.  So I disagree with removing that until
there's more stability. I just had to repair 15 sensors this weekend.  I'd
really rather it stayed unless it actually causes trouble.

On Wed, Feb 7, 2018 at 10:16 PM, Keith G <notifications@github.com> wrote:

> @alecm <https://github.com/alecm> - by design, the lastCheckin event only
> happens when there's a report message from the sensor (i.e.., battery,
> temp, humidity, pressure, motion detected/no motion , lux, open/close state
> change, wet/dry state change, and button pushed.)
>
> In my updated code for the T/H sensor, I turned off the display of
> lastCheckin event messages for *everything* except rmessages that can't
> be parsed. And I have not seen a single checkin event displayed since then.
> On my new motion detector, when I'm not at home and no motion is detected,
> every single lastCheckin happens at the same time as a battery report. Same
> thing for all my door/window sensors and my leak sensor.
>
> So, these sensors *do not* actually send a real check-in message. Users
> should know that during times of inactivity, battery reports are the best
> indicator that the sensor is still working and connected. I have seen a few
> instances of skipped battery reporting, but in those cases there's no
> checkin message. This means all the checkin message is doing is cluttering
> up the events log as seen in the "Recently" tab of the St mobile app.
>
> However, I think having the last checkin date/time displayed in a tile is
> great, because if it's more than two hours ago, you need to see what's
> going on. So all types of reports in the new T/H sensor code still log an
> event, it's just not displayed. You can still find the full list of
> lastCheckin events by looking at the device in IDE, and click on
> lastCheckin in the Current State section. And these will of course all
> correspond to other report events.
>
> I also believe the Last Checkin tile should be renamed to something like
> "Last report received" because that is more representative of what it's
> displaying.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bspranger/Xiaomi/pull/83#issuecomment-363989733>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5P1_E59kht1aPafvxeRIFmO7Yyh4OTks5tSmcagaJpZM4R8tfQ>
> .
>



-- 
Alec McLure, MPH, RHIA, CCS-P
*Cell:* +1 401-663-1316
*Email*: alecdjm@gmail.com

veeceeoh commented:
I think I actually didn't explain well enough. :smile:

What you should be looking for is battery reports, roughly every hour. That is the only consistent form of "check in" that Xiaomi sensors provide.

They do not perform any kind of network check in that results in a message sent through to the device handler. So the lastCheckin events are a fallacy. The code was written to log a lastCheckin event along with any other event. In other words, they are a synthetic and a completely unnecessary duplication of actual real event messages. They won't help in troubleshooting or anything really. 

Again, it's battery reports that people should be watching for. Leaving lastCheckin in the code just produces a duplicate event that rides along with the battery report (or any other kind of report, as I mentioned above). It works that way because the device handler code is set to do that.

Does that make sense?

alecm commented:
Yup - I'm cool, @veeceeoh  - as long as I'm getting my 55 minute fix (for the Motion Sensors) - whether it's battery or "check in" - I'm good to go - thanks!   (others may struggle with that, though - so we should really do release notes to the group when this is finally pushed to explain that -)

veeceeoh commented:
So is this what you see when your motion sensor is idle for hours?

![img_2191](https://user-images.githubusercontent.com/33558908/35957296-c3d85b36-0c4f-11e8-9682-15c213d20168.PNG)

Every one of those lastCheckin messages comes at the same time as a battery report. Remove the superfluous lastCheckin events and now you can see twice as many of the "real" check-in events, when the battery report is sent!

On the "Recently" tab for my Door/Window Sensors, I can scroll through for days and days, and see the battery reports every 50-60 minutes. That's all that's needed to know they're still connected.

tmleafs commented:
@veeceeoh I tested adding selection titles like the Docs say you can for apps and it didn't do a thing so might as well remove them

tmleafs commented:
@veeceeoh have applied your version 4 changes to all the other DH handlers I have done the basics for the original temp mainly just adding comments but will leave the rest to you

I have removed sections and replaced them with comments instead

alecm commented:
@tmleafs did you mean you deleted "section" titles?   Which makes sense if they don't work.  Also, nice job adding some cleaner commentsto the drivers!

alecm commented:
@veeceeoh  No worries - you've convinced me - I agree.  Think I was reacting from days when battery reporting still hadn't been nailed down.  Lets just make sure the group gets it when we write it up.

alecm commented:
oops sorry - closed by mistake - reopened again - guessing didn't cause any problems. I won't be able to do any testing today, I'm afraid - heavy day at work.

tmleafs commented:
@alecm we had sections like `section() { }` but in the docs its says you can so this `section("Date & TIme") { }` but it didnt work proving section doesn't to anything in device handlers
see more[ here ](http://docs.smartthings.com/en/latest/smartapp-developers-guide/preferences-and-settings.html)

alecm commented:
@tmleafs Looks like section titles works in SmartApps but not DTHs - makes sense to remove.

veeceeoh commented:
@tmleafs - I'm working on the new code for the "original" Temp/Humidity sensor already. It's been really busy at work, doing overtime, so perhaps I can finish that tonight.

Thanks for all the updates in your Pull Req - you are very thorough!!

veeceeoh commented:
I have completed the update to the "original" Xiaomi T/H sensor DTH code. I hope it's okay that I merged it as a commit on this Pull Request to avoid any conflicts.

Since the all the functions (sans pressure reporting) are identical to the Aqara T/H sensor, I was able to test the code by changing the device handler assigned to my Aqara sensor. It's all working perfectly, but of course I can't test the pairing process, because the fingerprint profileId is not the same.

One thing that surprised me a bit was that the Last Checkin tile was missing from this DTH. Although I've removed the event reporting of lastCheckin, I still feel having the tile is helpful. It's a way to know when the sensor last sent a report to the ST hub. So I put it in, of course.

Without the Pressure tile, the smaller tile display looks a bit "one sided". I'm note sure how this could be improved upon except for using "spacer" tiles to center everything. Something to consider.

It would be great if anyone with an "original" T/H sensor who doesn't mind removing it and then re-pairing could check if that all works as expected.

veeceeoh commented:
I forgot that @tmleafs had changed the "Last Checkin" tile text to "Last Event". So I just changed that in the updated "original" T/H -and Aqara T/H code.

veeceeoh commented:
I applied @tmleafs standardization code updates to all the devices I have beside my Aqara T/H sensor, and it all looks good.

However, with my Aqara Motion Detector, I have noticed the preference for resetting to "No Motion" after X number of seconds is not working. I think I've seen posts about that recently on the ST Community Forums, so it would be a good idea to get that fixed (also on the DTH for the "original" Motion Detector, if applicable) before the v1.0 release of everything. It's late, so I won't have a look until tomorrow, maybe after work.

Another thought on the tiles - where applicable, the Last Event / Battery Changed tiles could be changed to 6 x 1 and the \n carriage return removed. Then it would look like they fill the extra space better. Also, for the T/H sensors, I am looking at adding code that another user came up with for displaying a "feels like" temperature value. I don't think that should hold up the overhauled T/H sensor code, but it would result in another 4x4 tile if the feature is added later.

tmleafs commented:
@veeceeoh 6 x 1 tiles display weird on android dunno about iOS thats why the buttons have 1x1 spacers to center the 4x1 instead. Would be interested what you see on iOS though

veeceeoh commented:
I have updated all DTHs with a change to getBatteryResult() that adds a safeguard against erroneous voltage readings, as identified in Issue #82. It's just a one line change, using my favorite thing, a ternary!

We don't have any other reports that this one user, so not a lot of information to go on, but it appears that in some cases either the sensor or the DTH's code drops the final zero of the 4 digit raw voltage number, for example 3190 comes in as 319. So thanks to a brilliant suggestion by @alecm, if the raw voltage integer is less than 1000, it will be divided by 100 instead of 1000. So in that example, we'd get a voltage of 3.19V to convert into a percentage.

Note that the update to the Aqara DTH was a commit on top of my code overhaul Pull Request, to avoid any conflict.

veeceeoh commented:
@tmleafs - If 6x1 tiles don't look good on the Android UI, then I say spacer tiles should be added to all DTHs that could benefit from them, just the same as with the button DTH.

Also, after having a careful look at both motion detector DTHs, I can see that they also need a code overhaul, in addition to figuring out the motion reset preference value being ignored. There are all kinds of functions and sections of code that aren't being used at all, are irrelevant or leave the code open to errors (such as remnant now date call that doesn't get passed on to the formatDate() function to correctly format it and avoid the no timezone issue.)

I can get that this weekend, but I only have the Aqara model to test with.

veeceeoh commented:
So after doing some testing of what happens when pairing a sensor, I decided to add the Battery Changed date reset to `configure()`, just to make sure the current date is auto-populated in the Battery Changed tile when a sensor is newly paired.

But just in case `configure()` might be called again later (for example, if the sensor connection is lost and then "found" again), I put in a conditional check to make sure that `batteryRuntime` is _null_. I also put this conditional in `installed()` just to be safe. In my testing with pairing a bunch of sensors, the Battery Changed date was auto-populated every time - as long as the "Add a Thing" method is used. I'm not even going to try the catchall method because that really isn't supported by ST.

During my testing, I discovered that the Aqara door/window sensor DTH was updating the "last opened" date in the main tile for both open _and close_ events. So I fixed that, and to both D/W sensor DTHs, I added `isStateChange: true` to the open/close event because it seems to help with responsiveness to the contacts been repeatedly opened and closed quickly. Finally, I thought it would be helpful to add an description to the open/close events when the override tile buttons are used. Now it says "was manually reset to open / closed" so it's clear that it wasn't an actual sensor event.

I have updated the battery changed reset routines on all of the DTHs in the three Pull Request branches currently active (`code-overhaul` for the Aqara T/H sensor, `veeceeoh-patch-1` for both motion detector sensors, and `standardization` for everything else.)

Next I'm going to merge `veeceeoh-patch-1` into `standardization`, and I actually think `standardization` should  be merged into `code-overhaul` so there's just the one Pull Request ready for a final check.

veeceeoh commented:
Okay, I merged the new motion detector DTH code into here.

So the next step is to merge this Pull Req into Code Overhaul, leaving us with one Pull Request, and a final check through to do.

@bspranger, @tmleafs, @alecm -  okay with that?

alecm commented:
Hiya- 

So,got my first battery report from my repaired original Xiaomi motion sensor a little while ago - 

![image](https://user-images.githubusercontent.com/3035095/36080506-346f42da-0f5f-11e8-89aa-ba626c5db589.png)

First time I've had one of these.   Again, it was my one original sensor, which I understand has always been miserable to decode...

Thinking we shouldn't hold off on pushing Aqara versions because of tracking down probs with the originals - at least w/ the motion sensors. 

(Yes, I have 15 very happy and effective Quirky Trippers in my system, but I'd never expect any additional support for them at this point - just a thought). 

Alec

veeceeoh commented:
The only way to figure it out is with the debug log messages at the time of the battery report.

It's good you posted a screenshot of the "original" motion detector DTH UI, because I see a mistake in my placement of the spacer tiles. I'll fix that now.

alecm commented:
Here you go - it's repeating on this one - looking at something else

![image](https://user-images.githubusercontent.com/3035095/36081990-5ad4e9e8-0f73-11e8-8819-61a59bd6d273.png)

veeceeoh commented:
Okay, I've figured it out.

The regular hourly battery reports are seen as a `catchall` message from all the Xiaomi sensors. Here's an example from my Aqara motion detector:

> Motion Sensor parsing: catchall: 0104 0000 01 01 0100 00 E667 00 01 115F 0A 01 **01FF**42210121**E50B**0328180421A81305210D00062401000000000A2100006410000B210000

The first 17 bytes are header information, providing things like the `profileId`, `senderId`, sourceEndpoint, `clusterId`, etc. The long string of bytes after that is the `data` portion of the message.

All the Xiaomi sensors send batter data from either clusterId `FF01` or `FF02`. In the example above `FF01` in boldface is that clusterId (delivered big-endian, so swap the positions of the two bytes). Then there some more bytes with what I guess you could call reporting message parameters. The two bytes with the raw battery voltage value are in a different spot of the data portion of the catchall message of each Xiaomi sensor, but generally they're 4 to 10 bytes after the `FF01` or `FF02` `clusterId` that is at the beginning of the catchall data string.

In this case the battery voltage value bytes are numbers 7 and 8 (boldfaced `E50B`, again big-endian, so reverse them, and you've got `0BE5` which converted to an integer is 3045, and divided by 1000 gives you 3.045V. So that's how the data is delivered.

The way the battery value is read in by most DTHs is using a `for()` loop method @bspranger came up with that steps through the catchall data string searching for the byte _before_ the two battery bytes, which is always `21`. Here's the specific code which does that:

```
for (int i = 0; i < (MsgLength-3); i++) {
    if (catchall.data.get(i) == 0x21) { // check the data ID and data type
        // next two bytes are the battery voltage
        resultMap = getBatteryResult((catchall.data.get(i+2)<<8) + catchall.data.get(i+1))
    }
    break
}
```

But there are two problems, one going on in the "original" motion detector DTH, and the other problem is in all DTHs.

The problem in the  "original" motion detector DTH is that the `break` command is missing, so the `for()` loop just keeps going through the entire string looking for every instance of a byte `21` (except for the last 3 bytes as specified by` i < (MsgLength-3)`.

In my example above it would use bytes `0000` for the battery voltage value, which are right after   the _second to last_ instance of a byte `21`, as boldfaced below:

01FF42210121E50B0328180421A81305210D00062401000000000A21**0000**6410000B210000

And in my test with my Aqara motion sensor using the "original" motion sensor DTH, that's exactly what happened. It saw the voltage as 0V and the percentage was -500%!

So, fine, we fix that by putting the `break` command back in where it's supposed to be. But that doesn't fix the second problem. With the `break` command, the loop will stop at the _first_ instance of the byte `21` that is found, and grab the two bytes after that to use for the voltage value.

Again, using my example, that means the bytes `2101` would get used - the equivalent of _8.449V_, because the for() loop starts its search for `21` at the first byte, and would find the first instance at byte 4, and grab bytes 5 & 6 (in reverse order) to calculate the voltage and the percentage. All of which would be incorrect.

So I will have to check [this web page on the ZiGate USB device website](https://zigate.fr/produits-xiaomi-compatibles-zigate/) that gives a wealth of technical information about ZigBee messages sent by Xiaomi devices and see which byte number the loop should start at to safely get it past any incorrect instances of the byte `21` as seen in my example. I'm going to guess starting at byte 5 or 6 will be safe.

The other way to handle it is to look at a catchall message sent by every single type of sensor, identify the two bytes with battery data, and then in the DTH just tell it exactly which two bytes to read, instead of using the loop.

Phew! Sorry for the very technical explanation. This needs to be figured out before all of the updated DTHs are released though.

veeceeoh commented:
@alecm - your screenshot just confirms what I said about problem #1. The bytes after the last instance of `21` are `40` and `00`, which reversed are equal to 64, which is less than 1000, so it got divided by 100 to get 0.64V!!

However, just for that particular data string, if we put the break command in where it belongs you would have seen the correct voltage because there aren't any instances of byte `21` before the correct one. So your actual raw voltage reading was `0BB3`, equal to 2995 divided by 1000 to get 2.995V.

alecm commented:
You're very impressive @veeceeoh Keith - I hadn't found that zigate page and I've been following the guy for a while (I speak fairly fluent French) .  Does this really mean we have to hold off on all the DTHs or just on the originals?

veeceeoh commented:
I would say hold off on all of them. I have no idea if those first 6-8 bytes past the initial `FF01` / `FF02` `clusterId` change from one battery report to the next.

And yes, I have also scoured the blog page of Frédéric Dubois for all the information I could get on Xiaomi devices. His information has been instrumental in my understanding how they work. Brilliant guy.

veeceeoh commented:
Okay, based on what Frédéric explains on the page with specifics about message from the Aqara T/H sensor, the first _four_ bytes of the catchall message should be ignored by the `for()` loop.

The SmartThings ZigBee parser scrambles things by reversing the -endian order of 16bit hex words, but using my catchall message example from before, here's what the first 4 bytes mean:

Original first 4 bytes of catchall message = `01FF4221`
Those six bytes with 16bit hex words -endian reversed = `FF012142`
`FF01` is Cluster ID
`21` is Attribute Size
`42` is Attribute Type (Character String)
Everything after that is the Attribute Data.

I've looked at logs for all of my devices, and starting the `for()` loop on byte 5 will work, but then the only type of "original" Xiaomi device I have is the door/window sensor. 

All the Aqara sensors I have (D/W, T/H, MD, Leak, & 2-switch button) have the battery data in bytes 7 & 8 of the catchall data, but for the "original" D/W sensor, it's found at bytes 9 & 10 (but all the preceding byte values never change in any of the logged catchall messages.) Either way, I'm convinced it's safe to ignore those first four bytes.

So, I'm going to add the break command to any of the DTHs that don't have it, and change the for() loop to start at byte 5 on all DTHs. Then I'll watch what happens in my logs. It would be useful to hear what happens in the logs of anyone's "original" T/H, motion, and button devices.

veeceeoh commented:
Well, something is not working correctly with the fix. Battery parse is coming up blank.

My eyeballs are going to fall out of my head, so I will have a fresh look tomorrow.

All DTHs are definitely **not** ready for release yet.... argh.

veeceeoh commented:
My eyeballs started working again, and I found my mistake: I put the break `command` in the wrong place.

Now working correctly in all DTHs after some very thorough testing with all my devices.

I also removed the divide by 100 method I originally devised to try to deal with the erroneous battery report as originally mentioned in Issue #82. The problem wasn't a zero being dropped, but instead an unforeseen situation with using the `for()` loop as I explained in my incredibly long post above.

Anyhow, the revised battery parse code in the `parseCatchAllMessage()` function is ready for prime time, as far as I'm concerned.

alecm commented:
OK - so I have 
- button original
- motion sensor original
- motion sensor Aqara
- temp original
- temp Aqara
(only device types I have)
running for now in your newest version

veeceeoh commented:
I have checked on every single one of my devices, and battery has been reporting correctly on all of them for nearly 12 hours now. So I think we're good on that.

I'm going to merge this into the code overhaul Pull Req.
